File: python/src/s2iam/https.py
"""HTTPS validation for authentication server URLs."""

from urllib.parse import urlparse


def validate_auth_server_url(raw_url: str, *, allow_http: bool = False) -> None:
    """Ensure the authentication server URL uses HTTPS unless allow_http is set."""
    parsed = urlparse(raw_url)
    scheme = parsed.scheme
    if scheme == "https":
        return
    if scheme == "http" and allow_http:
        return
    if scheme == "http":
        raise ValueError(
            "authentication server URL must use HTTPS; pass allow_http=True for testing"
        )
    raise ValueError(f"authentication server URL must use HTTPS (got scheme {scheme!r})")

File: go/s2iam/https.go
package s2iam

import (
	"net/url"

	"github.com/memsql/errors"
	"github.com/singlestore-labs/codegate"
)

// TODO: INFOSEC-3101 remove this code gate once HTTPS enforcement is stable at production
var gateS2IAMRequireHTTPS = codegate.New("S2IAMRequireHTTPS")

func validateAuthServerURL(rawURL string, allowHTTP bool) (*url.URL, error) {
	uri, err := url.Parse(rawURL)
	if err != nil {
		return nil, errors.Errorf("invalid server URL: %w", err)
	}
	if !gateS2IAMRequireHTTPS.Enabled() {
		return uri, nil
	}
	switch uri.Scheme {
	case "https":
		return uri, nil
	case "http":
		if allowHTTP {
			return uri, nil
		}
		return nil, errors.New("authentication server URL must use HTTPS; use WithAllowHTTP() for testing")
	default:
		return nil, errors.Errorf("authentication server URL must use HTTPS (got scheme %q)", uri.Scheme)
	}
}
